ingestion-beam/src/test/java/com/mozilla/telemetry/util/DynamicPathTemplateTest.java

```diff
@@ -88,6 +92,20 @@ public void testExtractValuesFrom() {
     assertThat(path.extractValuesFrom(null), is(ImmutableList.of("b", "f")));
   }
 
+  @Test
+  public void testExtractValuesFromWithQualifiedName() {
+    final DynamicPathTemplate path = new DynamicPathTemplate("tmp/${bar:-b}/test", true);
+    final ImmutableMap<String, String> attributes = ImmutableMap.of("bar", "hi", "unused", "blah",
+        "document_namespace", "x", "document_type", "y", "document_version", "z");
+    final ImmutableMap<String, String> partialQualifiedNameAttributes = ImmutableMap
+        .of("document_type", "z");
+
+    assertThat(path.extractValuesFrom(attributes), is(ImmutableList.of("x", "y", "z", "hi")));
+    assertThat(path.extractValuesFrom(null), is(ImmutableList.of("$", "$", "$", "b")));
+    assertThat(path.extractValuesFrom(partialQualifiedNameAttributes),
+        is(ImmutableList.of("$", "z", "$", "b")));
+  }
+
```

Comment:
Can you add a test case that more closely matches the reason the new methods are necessary for the Avro sink? If we wrote to a directory path that includes namespace, doctype, and version and thus placeholders like `${document_namespace:-foo}` appeared in the template, would these new methods be unnecessary? Could we document that paths must include those three attributes and raise an exception if that is violated?

It feels strange to have this use case-specific logic baked into this class. Maybe this would feel more natural as a subclass of DynamicPathTemplate rather than using a boolean constructor param to toggle on this functionality.

Reply:
Can do. The core bit is being able to reconstruct a Map<PlaceholderName, PlaceholderValue> from the List<PlaceholderValue>, which is only possible if the placeholder names are in scope and the names contain the three required attributes. I'll rework it to see if I can make it flow more naturally.